src/main/java/com/redhat/devtools/lsp4ij/features/selectionRange/LSPExtendWordSelectionHandler.java

```diff
@@ -0,0 +1,194 @@
+ *
+ * Contributors:
+ * Red Hat, Inc. - initial API and implementation
+ ******************************************************************************/
+package com.redhat.devtools.lsp4ij.features.selectionRange;
+
+import com.intellij.codeInsight.editorActions.ExtendWordSelectionHandler;
+import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.editor.Editor;
+import com.intellij.openapi.progress.ProcessCanceledException;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.util.TextRange;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiFile;
+import com.intellij.util.containers.ContainerUtil;
+import com.redhat.devtools.lsp4ij.LSPFileSupport;
+import com.redhat.devtools.lsp4ij.LSPIJUtils;
+import com.redhat.devtools.lsp4ij.LanguageServerItem;
+import com.redhat.devtools.lsp4ij.LanguageServiceAccessor;
+import com.redhat.devtools.lsp4ij.client.ExecuteLSPFeatureStatus;
+import com.redhat.devtools.lsp4ij.client.features.AbstractLSPDocumentFeature;
+import com.redhat.devtools.lsp4ij.client.features.LSPClientFeatures;
+import com.redhat.devtools.lsp4ij.client.indexing.ProjectIndexingManager;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.Range;
+import org.eclipse.lsp4j.SelectionRange;
+import org.eclipse.lsp4j.TextDocumentIdentifier;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+
+import static com.intellij.codeInsight.editorActions.ExtendWordSelectionHandlerBase.expandToWholeLinesWithBlanks;
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.isDoneNormally;
+import static com.redhat.devtools.lsp4ij.internal.CompletableFutures.waitUntilDone;
+
+/**
+ * Implementation of the IDE's extendWordSelectionHandler EP for LSP4IJ files against textDocument/selectionRange.
+ */
+public class LSPExtendWordSelectionHandler implements ExtendWordSelectionHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LSPExtendWordSelectionHandler.class);
+
+    @Override
+    public boolean canSelect(@NotNull PsiElement element) {
+        if (!element.isValid()) {
+            return false;
+        }
+
+        PsiFile file = element.getContainingFile();
+        if ((file == null) || !file.isValid()) {
+            return false;
+        }
+
+        // Only if textDocument/selectionRange is supported for the file
+        return isSupported(file, LSPClientFeatures::getSelectionRangeFeature);
+    }
+
+    // TODO: This seems so incredibly boiler-plate that it should already exist somewhere common
+    private static boolean isSupported(@NotNull PsiFile file,
```

Comment:
Is there not already something that does this in an encapsulated manner like this? If not, should this be moved into a common location?

Reply:
The short answer is https://github.com/redhat-developer/lsp4ij/pull/669/files#r1877498632

Here a detailled explanation: when you need to check that an IJ feature is enabled (ex : to enable/disable Implementations menu item,  selectionrange like you need for your PR) it MUST be very fast otherwise IJ will freeze. 

You must never wait for starting language servers which can takes some times (some language servers can take 15sec to start it), so IJ could freeze during 15 sec.

Using document matcher requires to open a read action. In Ij Quarkus we use that to associate our Quarkus Language server to Java files only when the project have some Java class in the classpath.

Using hasAny is very fast. It checks:

 * if there is one or more language server which are associated to the file
 * if there is one language server has thecapability but ONLY when language server is started.

It means that if you execute Ctrl+W and language servers is starting, the action will be disable but I think it is better to loose the Ctrl+W feature and avoid freezing IJ.